src/util/metrics.cpp

```diff
@@ -0,0 +1,44 @@
+
+#include <stdint.h>
+
+// TODO: Clean up Arduino code from this file
+#include <Arduino.h>
+
+#include "../util/utils.h"
+#include "../util/metrics.h"
+
+void metricsReset(struct metrics* metrics)
+{
+  metrics->count = 0;
+  metrics->average = 0;
+  metrics->maximum = 0;
+  metrics->minimum = UINT32_MAX;
+  return;
+}
+
+void metricsStart(struct metrics* metrics)
+{
+  metrics->current = (uint32_t) micros();
+  return;
+}
+
+void metricsStop(struct metrics* metrics)
+{
+  uint32_t time = ((uint32_t) micros()) - metrics->current;
+  metrics->count++;
+  
+  // Handle rollover of sum (which should always rollover faster than sum
+  if (metrics->sum + time < metrics->sum) {
+    // TODO: Find a better solution
+    //       For now, set sum to a reasonable multiplier of the average so the current
+    //       time can continue to be rolled in without dominating the average
```

Comment:
One problem here is that the average is dependent upon how long we have been running/how long it has been since the last overflow, which is different for each set of metrics. What if we implement the average in a similar way to the sensor decimation filters? We could have a "running average" for the last n seconds of runtime, and even multiple averages (e.g. 1s, 10s, 60s).

Reply:
Yes, we can explore other averaging. This one is setup to take the average between calls to `metricsReset` as best as is possible. The overflow mechanism here is actually somewhat like the sensor decimation filter, though not implemented the same way. I had tried implementing averaging without `sum` (`average = (average * (count - 1) + current) / count`), but that approach doesnt work over many iterations as it trends towards 0 over time. We can experiment with it though trying to do averaging over time windows, while useful, might be hard to implement as we dont know when each sample came in to remove it. Certainly we can do the last n samples, but its harder to do last n seconds since this isnt a periodic function.

Reply:
And true about performance. In general, Im not trying to optimize too much until we find we need it -- sometimes is nice and easy to optimize so I go ahead and do it, but when its less obvious and requires a bit of thought, I usually leave that for future work if we find we need it.